tree/dataframe/inc/ROOT/RDataFrame.hxx

```diff
@@ -47,6 +48,10 @@ public:
    RDataFrame(TTree &tree, const ColumnNames_t &defaultBranches = {});
    RDataFrame(ULong64_t numEntries);
    RDataFrame(std::unique_ptr<ROOT::RDF::RDataSource>, const ColumnNames_t &defaultBranches = {});
+
+   /// \cond HIDDEN_SYMBOLS
+   RDataFrame(ROOT::Internal::RDF::RDatasetSpec spec);
```

Comment:
can this be private? we don't want user code that uses RDatasetSpec to work in 6.26, because it would do wrong things in multi-thread runs. we should give access just to distRDF.

Reply:
yes that's one way to give friend access from Python :D

Reply:
What we were currently trying:
```diff
diff --git a/tree/dataframe/inc/ROOT/RDataFrame.hxx b/tree/dataframe/inc/ROOT/RDataFrame.hxx
index d27b8ddcf7..904164817f 100644
--- a/tree/dataframe/inc/ROOT/RDataFrame.hxx
+++ b/tree/dataframe/inc/ROOT/RDataFrame.hxx
@@ -32,6 +32,15 @@ class TDirectory;
 class TTree;
 
 namespace ROOT {
+
+class RDataFrame;
+
+namespace Internal {
+namespace RDF {
+ROOT::RDataFrame MakeDataFrameFromSpec(RDatasetSpec spec);
+}
+}
+
 namespace RDF {
 class RDataSource;
 }
@@ -48,12 +57,20 @@ public:
    RDataFrame(TTree &tree, const ColumnNames_t &defaultBranches = {});
    RDataFrame(ULong64_t numEntries);
    RDataFrame(std::unique_ptr<ROOT::RDF::RDataSource>, const ColumnNames_t &defaultBranches = {});
+   friend RDataFrame ROOT::Internal::RDF::MakeDataFrameFromSpec(ROOT::Internal::RDF::RDatasetSpec spec);
 
-   /// \cond HIDDEN_SYMBOLS
+private:
    RDataFrame(ROOT::Internal::RDF::RDatasetSpec spec);
-   /// \endcond
 };
 
+//namespace Internal {
+//namespace RDF {
+//ROOT::RDataFrame MakeDataFrameFromSpec(const ROOT::Internal::RDF::RDatasetSpec &spec) {
+//   return ROOT::RDataFrame(spec);
+//}
+//}
+//}
+
 } // ns ROOT
```

And also adding the files `tree/dataframe/inc/ROOT/RDF/InternalUtils.hxx` and `tree/dataframe/src/RDFInternalUtils.cxx` respectively:
```cpp
#include "ROOT/RDF/RDatasetSpec.hxx"
#include "ROOT/RDataFrame.hxx"

namespace ROOT {
namespace Internal {
namespace RDF {

ROOT::RDataFrame MakeDataFrameFromSpec(RDatasetSpec spec);

} // namespace RDF
} // namespace Internal
} // namespace ROOT
```
and
```cpp
#include "ROOT/RDF/InternalUtils.hxx"

namespace ROOT {
namespace Internal {
namespace RDF {

ROOT::RDataFrame MakeDataFrameFromSpec(const RDatasetSpec &spec)
{
   return ROOT::RDataFrame(spec);
}

} // namespace RDF
} // namespace Internal
} // namespace ROOT
```

Reply:
your declaration does not match the definition so the compiler does not find a definition for your declaration :)

Reply:
@eguiraud I am looking at that again and I still cannot see it!